### PR TITLE
KAFKA-12950 Replace EasyMock and PowerMock with Mockito for KafkaStream

### PR DESCRIPTION
**Pull request by wycccccc**
Development of EasyMock and PowerMock has stagnated while Mockito continues to be actively developed. With the new Java cadence, it's a problem to depend on libraries that do bytecode generation and are not actively maintained. In addition, Mockito is also easier to [use.KAFKA-7438](https://issues.apache.org/jira/browse/KAFKA-12950)

`build.gradle `will update after other related issues are merged, avoid conflicts.Minor changes to the source code.I think this is inevitable.If there is a better solution to avoid changes, suggestions are welcome.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 1]
```diff
@@ -939,6 +939,11 @@ private StreamThread createAndAddStreamThread(final long cacheSizePerThread, fin
         return streamThread;
     }
 
+    //For KafkaStreamsTest
```
Comment (chia7712) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Why? Could you please add more details?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 2]
```diff
@@ -246,10 +246,15 @@ State setState(final State newState) {
 
     public boolean isRunning() {
         synchronized (stateLock) {
-            return state.isAlive();
+            return getIsAlive();
         }
     }
 
+    //For KafkaStreamTest
+    public boolean getIsAlive() {
```
Comment (chia7712) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: this naming is a bit weird to me. How about `isStateAlive`?

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 3]
```diff
@@ -96,13 +94,7 @@
 import static org.apache.kafka.streams.integration.utils.IntegrationTestUtils.safeUniqueTestName;
 import static org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitForApplicationState;
 
-import static org.easymock.EasyMock.anyBoolean;
 import static org.apache.kafka.test.TestUtils.waitForCondition;
-import static org.easymock.EasyMock.anyInt;
```
Comment (chia7712) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:100]: Is `easymock` still used by streams module? If not, could you remove it from `build.gradle`?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 4]
```diff
@@ -246,10 +246,15 @@ State setState(final State newState) {
 
     public boolean isRunning() {
         synchronized (stateLock) {
-            return state.isAlive();
+            return getIsAlive();
         }
     }
 
+    //For KafkaStreamTest
```
Comment (chia7712) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: ditto. Could you add more description?

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 5]
```diff
@@ -182,152 +192,138 @@ public void before() throws Exception {
...
-                StreamThread.State.PARTITIONS_REVOKED,
-                StreamThread.State.STARTING);
-            threadStatelistenerCapture.getValue().onChange(thread,
-                StreamThread.State.PARTITIONS_ASSIGNED,
-                StreamThread.State.PARTITIONS_REVOKED);
-            threadStatelistenerCapture.getValue().onChange(thread,
-                StreamThread.State.RUNNING,
-                StreamThread.State.PARTITIONS_ASSIGNED);
+            threadStateListenerCapture.getValue().onChange(thread,
+                    StreamThread.State.STARTING,
```
Comment (chia7712) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: redundant indent?

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 6]
```diff
@@ -61,18 +59,18 @@
...
-import org.easymock.Capture;
-import org.easymock.EasyMock;
 import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
+import org.junit.After;
 import org.junit.rules.TestName;
-import org.junit.runner.RunWith;
-import org.powermock.api.easymock.PowerMock;
```
Comment (chia7712) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:72]: Is `powermock` still used by streams module? If not, could you remove it from build.gradle?

**Review by chia7712 — COMMENTED** [review]
@wycccccc thanks for this nice patch. a couple of comments below.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 2 above)
Comment (wycccccc) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: copy that

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 1 above)
Comment (wycccccc) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: I have update code, thanks for all your comments.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java (hunk 3 above)
Comment (wycccccc) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:100]: I am worried that the current changes for `build,gradle` will cause conflicts. I am going to prepare a separate PR after all related issues are merged.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java (hunk 6 above)
Comment (wycccccc) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:72]: ditto

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 7]
```diff
@@ -939,7 +939,7 @@ private StreamThread createAndAddStreamThread(final long cacheSizePerThread, fin
         return streamThread;
     }
 
-    //For KafkaStreamsTest
+    //Ensure Mockito stub construct with capture argument for KafkaStreamsTest.
```
Comment (tang7526) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Could you add a space between `//` and `Ensure`?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 8]
```diff
@@ -246,12 +246,12 @@ State setState(final State newState) {
...
     public boolean isRunning() {
         synchronized (stateLock) {
-            return getIsAlive();
+            return isStateAlive();
         }
     }
 
-    //For KafkaStreamTest
-    public boolean getIsAlive() {
+    //Ensure Mockito can stub method for KafkaStreamTest.
```
Comment (tang7526) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Could you add a space between // and Ensure?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 7 above)
Comment (wycccccc) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Have changed.

**Comment by ijuma** [comment]
@wycccccc Can you please fix the conflict? @chia7712 How do you feel about this PR, is it close to being ready to be merged?

**Comment by chia7712** [comment]
> How do you feel about this PR, is it close to being ready to be merged?

I need to take another look after those conflicts are fixed.

**Comment by wycccccc** [comment]
@chia7712 The conflict has been resolved, let me know if you have any situation.

**Comment by ijuma** [comment]
Btw, this is the only Streams test that is disabled with Java 16 and newer:
https://github.com/apache/kafka/blob/trunk/build.gradle#L389

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 9]
```diff
@@ -111,9 +104,24 @@
...
+import static org.mockito.Mockito.isA;
+import static org.mockito.Mockito.eq;
+import static org.mockito.Mockito.anyInt;
+import static org.mockito.Mockito.anyLong;
+import static org.mockito.Mockito.withSettings;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.MockitoAnnotations.openMocks;
 
-@RunWith(PowerMockRunner.class)
```
Comment (ijuma) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:115]: Are there other Streams tests using `PowerMock`? If not, we should also update the module build dependencies.

Comment (wycccccc) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:115]: I remember that I have modified all the streams and I have updated the module build dependencies.

**Comment by ijuma** [comment]
@chia7712 Conflicts are resolved. :)

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 10]
```diff
@@ -977,6 +977,11 @@ private StreamThread createAndAddStreamThread(final long cacheSizePerThread, fin
         return streamThread;
     }
 
+    // Ensure Mockito stub construct with capture argument for KafkaStreamsTest.
+    public static Metrics createThisMetrics(final MetricConfig metricConfig, final List<MetricsReporter> reporters, final Time time, final MetricsContext metricsContext) {
```
Comment (chia7712) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:981]: Why need the word "this"?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 11]
```diff
@@ -246,10 +246,15 @@ State setState(final State newState) {
 
     public boolean isRunning() {
         synchronized (stateLock) {
-            return state.isAlive();
+            return isStateAlive();
         }
     }
 
+    // Ensure Mockito can stub method for KafkaStreamTest.
+    public boolean isStateAlive() {
```
Comment (chia7712) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:254]: Personally, it is an anti-pattern. There are two Public methods returning same "value" but one of them is holding lock. It is hard to use them correctly. At any rate, this is unrelated to this PR.

File: build.gradle [hunk 12]
```diff
@@ -1751,8 +1749,6 @@ project(':streams') {
     testImplementation libs.junitJupiterApi
     testImplementation libs.junitVintageEngine
     testImplementation libs.easymock
-    testImplementation libs.powermockJunit4
```
Comment (chia7712) [inline build.gradle:1754]: `TableSourceNodeTest` is still using powermock

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 13]
```diff
@@ -335,41 +332,37 @@ private void prepareStreamThread(final StreamThread thread,
...
-        EasyMock.expectLastCall().anyTimes();
-        thread.requestLeaveGroupDuringShutdown();
-        EasyMock.expectLastCall().anyTimes();
-        EasyMock.expect(thread.getName()).andStubReturn("processId-StreamThread-" + threadId);
-        thread.shutdown();
-        EasyMock.expectLastCall().andAnswer(() -> {
+        ));
+        when(thread.waitOnThreadState(isA(StreamThread.State.class), anyLong())).thenReturn(true);
+        when(thread.isStateAlive()).thenReturn(true);
+        verify(thread, atMostOnce()).isStateAlive();
```
Comment (chia7712) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:338]: It seems to me this check is no-op since the thread is not ready to be used. Hence, the execution times is always 0.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 14]
```diff
@@ -335,41 +332,37 @@ private void prepareStreamThread(final StreamThread thread,
...
+            threadStateListenerCapture.getValue().onChange(thread, StreamThread.State.DEAD, StreamThread.State.PENDING_SHUTDOWN);
             return null;
-        }).anyTimes();
-        EasyMock.expect(thread.isRunning()).andReturn(state.get() == StreamThread.State.RUNNING).anyTimes();
-        thread.join();
+        }).when(thread).shutdown();
+
+        when(thread.isRunning()).thenReturn(state.get() == StreamThread.State.RUNNING);
+
         if (terminable) {
```
Comment (chia7712) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:356]: If you don't want to check the `join` method, please remove this empty body.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 15]
```diff
@@ -903,20 +897,15 @@ public void shouldTriggerRecordingOfRocksDBMetricsIfRecordingLevelIsDebug() {
...
-        final ScheduledExecutorService cleanupSchedule = EasyMock.niceMock(ScheduledExecutorService.class);
-        final ScheduledExecutorService rocksDBMetricsRecordingTriggerThread =
-            EasyMock.mock(ScheduledExecutorService.class);
-        EasyMock.expect(Executors.newSingleThreadScheduledExecutor(
-            anyObject(ThreadFactory.class)
-        )).andReturn(cleanupSchedule);
-        PowerMock.replay(Executors.class, rocksDBMetricsRecordingTriggerThread, cleanupSchedule);
+        final MockedStatic<Executors> executorsMockedStatic = mockStatic(Executors.class);
+        final ScheduledExecutorService cleanupSchedule = mock(ScheduledExecutorService.class, withSettings().lenient());
+        final ScheduledExecutorService rocksDBMetricsRecordingTriggerThread = mock(ScheduledExecutorService.class);
```
Comment (chia7712) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:906]: please make sure `rocksDBMetricsRecordingTriggerThread` is not used to create thread in this test

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 16]
```diff
@@ -874,26 +868,26 @@ public void shouldNotBlockInCloseForZeroDuration() {
...
-        EasyMock.expect(rocksDBMetricsRecordingTriggerThread.shutdownNow()).andReturn(null);
-        PowerMock.replay(Executors.class);
-        PowerMock.replay(rocksDBMetricsRecordingTriggerThread);
-        PowerMock.replay(cleanupSchedule);
+            mock(ScheduledExecutorService.class);
+
+        executorsMockedStatic.when(() -> Executors.newSingleThreadScheduledExecutor(
+            any(ThreadFactory.class)
+        )).thenReturn(cleanupSchedule);
+        executorsMockedStatic.when(() -> Executors.newSingleThreadScheduledExecutor(
```
Comment (chia7712) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:879]: The different `ThreadFactory` should get different `ScheduledExecutorService` (see production code `KafkaStreams`). That was a bug in previous test. However, it was fine since all this test want to do is to make sure `rocksDBMetricsRecordingTriggerThread` is executed.

Could you follow the production code to return different `ScheduledExecutorService`? or add comments to say "this is fine because ..."

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 17]
```diff
@@ -903,20 +897,15 @@ public void shouldTriggerRecordingOfRocksDBMetricsIfRecordingLevelIsDebug() {
             streams.start();
         }
 
-        PowerMock.verify(Executors.class);
```
Comment (chia7712) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:906]: please add `verify` to make sure `rocksDBMetricsRecordingTriggerThread` is used to create thread in this test.

**Review by chia7712 — COMMENTED** [review]
@wycccccc thanks for this patch. a couple of comments below. please take a look.

**Comment by divijvaidya** [comment]
@wycccccc are you planning to work further on this PR?
